### PR TITLE
Adding arguments to RouteSettings and Navigator.pushNamed

### PR DESCRIPTION
**Pull request by gspencergoog**
This addresses issue #6225.

I've added a `Map<String, dynamic>` optional argument to RoutSettings, and an associated simple test.

One thing I've noticed while doing this is that the named routes seem to be embedded in the Navigator, instead of being a separate class, which would seem to be more generic, and it may mislead people into thinking that they need to use named routes instead of just creating their own Route<T> subclass.  It might be good to factor that out.

### CODE REVIEW

**Comment by Hixie** [comment]
Why a Map rather than anything else?

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -558,8 +566,9 @@ class Navigator extends StatefulWidget {
   /// ```dart
   /// Navigator.pushNamed(context, '/nyc/1776');
   /// ```
-  static Future<dynamic> pushNamed(BuildContext context, String routeName) {
-    return Navigator.of(context).pushNamed(routeName);
+  static Future<dynamic> pushNamed(BuildContext context, String routeName,
+      {Map<String, dynamic> arguments}) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: please follow flutter style (see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)

Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: (this applies across the patch)

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -224,6 +227,11 @@ class RouteSettings {
   /// If null, the route is anonymous.
   final String name;
 
+  /// The arguments map that this route was created with.
+  ///
+  /// If null, there were no arguments given for this route.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: This should have a few paragraphs detailing the motivation for this feature, e.g. talking about how if the app is handling a URL, the path part of the URL will become the name of the route, and the query arguments will end up as a `Map<String, List<String>>` map in the arguments, talking about why it's a list (query arguments aren't guaranteed to be unique), about whether the arguments are URL-decoded (presumably they are), etc.

**Comment by gspencergoog** [comment]
I went with Map because you were dubious about dynamic, and Map seemed to match the use case (adding arguments a la query parameters).  I'm happy to switch it to dynamic if you think that's OK.  It *would* be more general purpose.

**Comment by Hixie** [comment]
Well there's still a dynamic here. :-)

One option is to use `T`, a generic type parameter, though I'm not 100% sure where we would slot it... presumably on `push()`, Route, and RouteSettings, though I'm not sure how good that would be.

Another option would be to have people subclass RouteSettings, and put whatever they want on there. So for example you could have a RouteSettingsWithQueryParameters for the use case we've talked about here.

If query parameters are the only use case we care about, then I suppose it would be `Map<String, List<String>>`.

IMHO `Map<String, dynamic>` isn't particularly better than `dynamic`.

**Review by gspencergoog — COMMENTED** [review]
What was the original use case for named routes as they exist?  It seems like it doesn't map well to external sources of routes (e.g. URLs).  I can understand wanting to be able to deep-link into an app, so having a "path" into that location makes sense, but it's not very clear to me who is generating those routes external to the app.  It seems like it works well as a convenience internal to the app, but as you pointed out in the issue, someone can just create a Route subclass if they want particular behavior.

So should I just scrap this PR until we know what the URL handling design is?  I can see creating a more involved PlatformUrlRoute class to handle the platform-initiated routes. It also seems like it would be good to factor out the named route support into a separate NamedRoute class (but leave the convenience functions on Navigator so as not to break code). Then it probably makes more sense to allow subclassing RouteSettings, since the RouteFactories take them.

It just feels like the Navigator has generic `Route<T>` support, and should stick to that, and that named routes should be able to implement their behavior in a separate class (and if not, then the generic support is lacking somehow).

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: I meant spaces inside the named arguments, though yeah, putting this across multiple lines might be better. Just a nit though.

**Comment by Hixie** [comment]
> What was the original use case for named routes as they exist? 

They come from an early time in the project where we thought we would be more URL-based than we are.

> So should I just scrap this PR until we know what the URL handling design is?

I would encourage you to try to figure out what that should be! :-)

It's possible the patch as-is is fine. It's possible we should have an entirely different approach. Or anywhere in between. What we need is a clear vision so we can defend the PR, whatever it ends up being.

**Pull request closed without merging**